### PR TITLE
[Wasm] Ensure monovm_initialize is called on startup and address other startup items

### LINKED ISSUE #38994: Assembly.GetEntryAssembly() returns null on Browser WASM
Assembly.GetEntryAssembly() returns null within Browser WASM. Also occurred when running `exePath_UserLevelNone` from `ConfigurationManagerTest.cs` on Browser. Similar to #21319 

Failure text:
```
[FAIL] System.Tests.AppDomainTests.FriendlyName
Assert.Equal() Failure
          ↓ (pos 0)
Expected: System.Runtime.Extensions.Tests
Actual:   DefaultDomain
          ↑ (pos 0)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

Related Code
```
public string FriendlyName
        {
            get
            {
                Assembly? assembly = Assembly.GetEntryAssembly();
                return assembly != null ? assembly.GetName().Name! : "DefaultDomain";
            }
        }
```

### LINKED ISSUE #39074: RuntimeInformation.RuntimeIdentifier unknown on Browser wasm
When running tests for `System.Runtime.InteropServices.RuntimeInformation.Tests`, this failure is encountered because `AppContext.GetData("RUNTIME_IDENTIFIER")` does not yet work for wasm, leading to `RuntimeInformation.RuntimeIdentifier` being unknown. It is typically set by the hosting app.
```
[FAIL] System.Runtime.InteropServices.RuntimeInformationTests.RuntimeIdentifierTests.VerifyOSRid
Assert.EndsWith() Failure:
Expected:    Wasm
Actual:   ···nown
at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

### LINKED ISSUE #39650: [Wasm] Assembly.Location does not return the full path
When calling `Assembly.Location`, it is expected to return the absolute path of the assembly.  In wasm, only the assembly name itself is returned.  

For example:

```
// say SomeClass exists in Foo.dll
 
string path = typeof(SomeClass).Assembly.Location;   
Assembly assembly = Assembly.LoadFile(path);
```
Path will return `Foo.dll` instead of the expected `/Foo.dll` and `Assembly.LoadFile` will throw an exception because `path` is not absolute.  

The primary problem is wasm is that we load the runtime assemblies from a byte array.  We take this approach to speed up the loading process and skip loading the assemblies into the VFS.

### PR DESCRIPTION
**Pull request by steveisok**
This PR addresses: 

- Calling monovm_initialize matches how other mono platforms init themselves (Fixes https://github.com/dotnet/runtime/issues/39074)
- Set the entry assembly (Fixes https://github.com/dotnet/runtime/issues/38994)
- Make sure the assembly location is set to empty for assemblies that are loaded from memory (Fixes https://github.com/dotnet/runtime/issues/39650)

### CODE REVIEW

File: src/mono/mono/metadata/appdomain.h [hunk 1]
```diff
@@ -106,6 +106,9 @@ mono_domain_foreach        (MonoDomainFunc func, void* user_data);
 MONO_API MONO_RT_EXTERNAL_ONLY MonoAssembly *
 mono_domain_assembly_open  (MonoDomain *domain, const char *name);
 
+MONO_API void
+mono_domain_ensure_entry_assembly (MonoDomain *domain, MonoAssembly *assembly);
```
Comment (CoffeeFlux) [inline src/mono/mono/metadata/appdomain.h:110]: I guess the wasm runtime can't use private headers? If so, this is fine.

**Review by CoffeeFlux — APPROVED** [review]

File: src/mono/mono/metadata/appdomain.h (hunk 1 above)
Comment (steveisok) [reply src/mono/mono/metadata/appdomain.h:110]: Yeah, I don't believe you can.  I get a compiler error.

File: src/mono/wasm/runtime/driver.c [hunk 2]
```diff
@@ -359,6 +359,14 @@ mono_wasm_load_runtime (const char *managed_path, int enable_debugging)
 	monoeg_g_setenv ("DOTNET_SYSTEM_GLOBALIZATION_INVARIANT", "1", 0);
 #endif
 
+	const char* appctx_keys[1];
+	appctx_keys[0] = "APP_CONTEXT_BASE_DIRECTORY";
```
[the remark this answers is not in the document]
Comment (steveisok) [reply src/mono/wasm/runtime/driver.c]: https://github.com/dotnet/runtime/blob/ecb231cb3d962d332d00344358b0559043d29f88/src/installer/corehost/cli/hostpolicy/coreclr.cpp#L137-L151

**Comment by safern** [comment]
The build failures on `Libraries Test Run release...` are a fallout from my change to move the test builds to the product build. Will investigate and put up a PR.

File: src/mono/wasm/runtime/driver.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (safern) [reply src/mono/wasm/runtime/driver.c]: So I guess right now we will just keep going to use the fallback to get the base directory from:

https://github.com/dotnet/runtime/blob/ecb231cb3d962d332d00344358b0559043d29f88/src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs#L12

**Review by safern — APPROVED** [review]

File: src/mono/wasm/runtime/driver.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (steveisok) [reply src/mono/wasm/runtime/driver.c]: Hmm...  I bet we can get rid of that file now.  I’ll try it out.

[the remark this answers is not in the document]
Comment (steveisok) [reply src/mono/wasm/runtime/driver.c]: So, the good news is that https://github.com/dotnet/runtime/blob/ecb231cb3d962d332d00344358b0559043d29f88/src/libraries/System.Private.CoreLib/src/System/AppContext.cs#L24 will return "/" that we set during initialize.  The bad news is that we still need to keep `AppContext.Browser.cs`.  Assembly.Location still does not return the full path for the runtime assemblies.

**Comment by steveisok** [comment]
@CoffeeFlux Easiest way I found to have Assembly.Location be an empty string when loaded from memory was to do this:
https://github.com/dotnet/runtime/blob/305ac4d2c5f73b7ed88c9c613c91dbccdb626de1/src/mono/mono/metadata/assembly.c#L2504-L2510

I've got almost all the tests that relied on Location passing except for two.  The tests that fail are Assembly.GetFile and Assembly.GetFiles.  In looking at https://github.com/dotnet/runtime/blob/fb6cc280611ec76fefd55b2f50c563e0fc6aa45c/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs#L411-L426, it'll fall to GetFilesInternal. That'll use image->filename, which is NULL ;-).

Do you have any suggestions on alternatives in the runtime?

File: src/mono/wasm/runtime/driver.c [hunk 3]
```diff
@@ -502,6 +512,8 @@ mono_wasm_assembly_get_entry_point (MonoAssembly *assembly)
 	uint32_t entry = mono_image_get_entry_point (image);
 	if (!entry)
 		return NULL;
+	
+	mono_domain_ensure_entry_assembly (root_domain, assembly);
```
Comment (lewing) [inline src/mono/wasm/runtime/driver.c:516]: This looks good other than it still won't get called on the blazor side.  Their entry point starts at: https://github.com/dotnet/aspnetcore/blob/master/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts#L54
I'm not sure what the best solution there is.

**Review by lewing — APPROVED** [review]

File: src/libraries/System.Reflection.MetadataLoadContext/tests/src/TestUtils/CoreMetadataAssemblyResolver.cs [hunk 4]
```diff
@@ -22,7 +23,17 @@ public override Assembly Resolve(MetadataLoadContext context, AssemblyName assem
                 name.Equals("System.Runtime.InteropServices", StringComparison.OrdinalIgnoreCase))
             {
                 if (_coreAssembly == null)
+                {
                     _coreAssembly = context.LoadFromStream(TestUtils.CreateStreamForCoreAssembly());
+                    //if (PlatformDetection.IsNotBrowser)
```
Comment (safern) [inline src/libraries/System.Reflection.MetadataLoadContext/tests/src/TestUtils/CoreMetadataAssemblyResolver.cs:28]: Why is this commented out? Do we need these comments anymore?

File: src/mono/wasm/runtime/driver.c (hunk 3 above)
Comment (lewing) [reply src/mono/wasm/runtime/driver.c:516]: For now it shouldn't block this.

File: src/libraries/System.Reflection.MetadataLoadContext/tests/src/TestUtils/TestUtils.cs [hunk 5]
```diff
@@ -253,7 +253,9 @@ public static void Dump(this CustomAttributeTypedArgument cta, int indent = 0)
 
         private static readonly Lazy<bool> s_useRuntimeTypesForTests = new Lazy<bool>(() =>
         {
-            if (File.Exists(Path.Combine(Path.GetDirectoryName(typeof(TestUtils).Assembly.Location), "UseRuntimeTypes.txt")))
+            var loc = AssemblyPathHelper.GetAssemblyLocation(typeof(TestUtils).Assembly);
+
+            if (File.Exists(Path.Combine(loc, "UseRuntimeTypes.txt")))
```
Comment (safern) [inline src/libraries/System.Reflection.MetadataLoadContext/tests/src/TestUtils/TestUtils.cs:258]: Should we just return false for Browser? `loc` in Browser will be the assembly name + ".dll"

File: src/libraries/System.Reflection/tests/AssemblyTests.cs [hunk 6]
```diff
@@ -158,17 +158,27 @@ public void GetFile()
             AssertExtensions.Throws<ArgumentException>(null, () => typeof(AssemblyTests).Assembly.GetFile(""));
             Assert.Null(typeof(AssemblyTests).Assembly.GetFile("NonExistentfile.dll"));
             Assert.NotNull(typeof(AssemblyTests).Assembly.GetFile("System.Reflection.Tests.dll"));
-            if (PlatformDetection.IsNotBrowser) // see https://github.com/dotnet/runtime/issues/39650
-                Assert.Equal(typeof(AssemblyTests).Assembly.GetFile("System.Reflection.Tests.dll").Name, typeof(AssemblyTests).Assembly.Location);
+
+            // For browser, Location is null.
+            string name = (PlatformDetection.IsNotBrowser) ?
```
Comment (safern) [inline src/libraries/System.Reflection/tests/AssemblyTests.cs:163]: Why not use the helper you added?

File: src/libraries/System.Reflection/tests/AssemblyTests.cs [hunk 7]
```diff
@@ -158,17 +158,27 @@ public void GetFile()
...
         [Fact]
         public void GetFiles()
         {
             Assert.NotNull(typeof(AssemblyTests).Assembly.GetFiles());
             Assert.Equal(1, typeof(AssemblyTests).Assembly.GetFiles().Length);
-            if (PlatformDetection.IsNotBrowser) // see https://github.com/dotnet/runtime/issues/39650
-                Assert.Equal(typeof(AssemblyTests).Assembly.GetFiles()[0].Name, typeof(AssemblyTests).Assembly.Location);
+
+            // For browser, Location is null.
+            string name = (PlatformDetection.IsNotBrowser) ?
```
Comment (safern) [inline src/libraries/System.Reflection/tests/AssemblyTests.cs:177]: Same question here?

File: src/libraries/System.Reflection.MetadataLoadContext/tests/src/TestUtils/CoreMetadataAssemblyResolver.cs (hunk 4 above)
Comment (steveisok) [reply src/libraries/System.Reflection.MetadataLoadContext/tests/src/TestUtils/CoreMetadataAssemblyResolver.cs:28]: I need to change bad habits 😜.

File: src/libraries/System.Runtime.Loader/tests/SatelliteAssemblies.cs [hunk 8]
```diff
@@ -18,15 +18,15 @@ public class SatelliteAssembliesTestsFixture
         public SatelliteAssembliesTestsFixture()
         {
             AssemblyLoadContext satelliteAssembliesTests = new AssemblyLoadContext("SatelliteAssembliesTests");
-            var satelliteAssembliesTestsPath = (PlatformDetection.IsBrowser ? "/" : "") + typeof(SatelliteAssembliesTests).Assembly.Location;
+            var satelliteAssembliesTestsPath = (PlatformDetection.IsBrowser ? "/" : "") + AssemblyPathHelper.GetAssemblyLocation(typeof(SatelliteAssembliesTests).Assembly);
```
Comment (akoeplinger) [inline src/libraries/System.Runtime.Loader/tests/SatelliteAssemblies.cs:21]:
```suggestion
            var satelliteAssembliesTestsPath = AssemblyPathHelper.GetAssemblyLocation(typeof(SatelliteAssembliesTests).Assembly);
```

File: src/libraries/System.Runtime.Loader/tests/SatelliteAssemblies.cs [hunk 9]
```diff
@@ -18,15 +18,15 @@ public class SatelliteAssembliesTestsFixture
...
             satelliteAssembliesTests.LoadFromAssemblyPath(satelliteAssembliesTestsPath);
 
             AssemblyLoadContext referencedClassLib = new AssemblyLoadContext("ReferencedClassLib");
-            var referencedClassLibPath = (PlatformDetection.IsBrowser ? "/" : "") + typeof(ReferencedClassLib.Program).Assembly.Location;
+            var referencedClassLibPath = (PlatformDetection.IsBrowser ? "/" : "") + AssemblyPathHelper.GetAssemblyLocation(typeof(ReferencedClassLib.Program).Assembly);
```
Comment (akoeplinger) [inline src/libraries/System.Runtime.Loader/tests/SatelliteAssemblies.cs:25]:
```suggestion
            var referencedClassLibPath = AssemblyPathHelper.GetAssemblyLocation(typeof(ReferencedClassLib.Program).Assembly);
```

File: src/libraries/System.Runtime.Loader/tests/SatelliteAssemblies.cs [hunk 10]
```diff
@@ -18,15 +18,15 @@ public class SatelliteAssembliesTestsFixture
...
             referencedClassLib.LoadFromAssemblyPath(referencedClassLibPath);
 
             AssemblyLoadContext referencedClassLibNeutralIsSatellite = new AssemblyLoadContext("ReferencedClassLibNeutralIsSatellite");
-            var referencedClassLibNeutralIsSatellitePath = (PlatformDetection.IsBrowser ? "/" : "") + typeof(ReferencedClassLibNeutralIsSatellite.Program).Assembly.Location;
+            var referencedClassLibNeutralIsSatellitePath = (PlatformDetection.IsBrowser ? "/" : "") + AssemblyPathHelper.GetAssemblyLocation(typeof(ReferencedClassLibNeutralIsSatellite.Program).Assembly);
```
Comment (akoeplinger) [inline src/libraries/System.Runtime.Loader/tests/SatelliteAssemblies.cs:29]:
```suggestion
            var referencedClassLibNeutralIsSatellitePath = AssemblyPathHelper.GetAssemblyLocation(typeof(ReferencedClassLibNeutralIsSatellite.Program).Assembly);
```

File: src/libraries/System.Runtime/tests/System/Reflection/ModuleTests.cs [hunk 11]
```diff
@@ -81,7 +81,15 @@ public void CustomAttributes()
         [Fact]
         public void FullyQualifiedName()
         {
-            Assert.Equal(Assembly.GetExecutingAssembly().Location, Module.FullyQualifiedName);
+            var loc = AssemblyPathHelper.GetAssemblyLocation(Assembly.GetExecutingAssembly());
+
+            // Browser will include the path (/), so strip it
```
Comment (akoeplinger) [inline src/libraries/System.Runtime/tests/System/Reflection/ModuleTests.cs:86]: hm so what does `Module.FullyQualifiedName` return for an in-memory assembly, does it also return an empty string like `Assembly.Location` did? it sounds like maybe we should fix/follow-up on this.

File: src/mono/mono/metadata/assembly.c [hunk 12]
```diff
@@ -2501,6 +2501,16 @@ mono_assembly_open_from_bundle (MonoAssemblyLoadContext *alc, const char *filena
 	for (i = 0; !image && bundles [i]; ++i) {
 		if (strcmp (bundles [i]->name, is_satellite ? filename : name) == 0) {
 			image = mono_image_open_from_data_internal (alc, (char*)bundles [i]->data, bundles [i]->size, FALSE, status, refonly, FALSE, name);
+#if TARGET_WASM
```
Comment (akoeplinger) [inline src/mono/mono/metadata/assembly.c:2504]: not sure if we need to keep this compatible but only do this on netcore?

```suggestion
#if defined(TARGET_WASM) && defined(ENABLE_NETCORE)
```

**Review by akoeplinger — COMMENTED** [review]
A few comments, but looks good in general

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**